---
title: "IonTorrent_ITS1_Analysis_FusionPrimerMethd"
author: "Daniel Gladish"
date: "1/19/2021"
output:
  pdf_document: default
  html_document: default
urlcolor: blue
header-includes: \usepackage{xcolor}
---

Set .libPaths() to the directory that houses the downloaded R packages.
```{r}
.libPaths(c("/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/R/x86_64-pc-linux-gnu-library/3.5" , .libPaths()))
```

```{r, global_options, eval=TRUE, echo=FALSE}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff = 80), tidy = TRUE, fig.align = 'center',
               cache = FALSE, collapse = TRUE, echo = FALSE, eval = FALSE, include = FALSE,
               message = FALSE, quietly = TRUE, results = 'hide', warn.conflicts = FALSE, 
               warning = FALSE)
```

**Using package `BiocManager` to install required packages:**
```{r, biocInstall, eval=TRUE, echo=TRUE, include=TRUE}
#Installing required packages
r <- getOption("repos")
r["CRAN"] <- "http://cran.us.r-project.org"
options(repos = r)

if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()

library("BiocManager")
.cran_packages <- c("data.table", "kableExtra", "knitr", "rprojroot")
.bioc_packages <- c("BiocStyle", "Biostrings", "dada2", "RAM")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  BiocManager::install(.bioc_packages[!.inst], ask = FALSE)
}
```
   
**Load packages into session, and print package versions:**
```{r, showBiocPackages, echo=TRUE, eval=TRUE, include=TRUE, results='hold'}
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
```
**Source our custom R scripts:**    
For this we will use the rprojroot package to set the directory structures. This will help us when finding our files to source functions. We specify ours is an RStudio project. The root object contains a function that will help us locate our package R files regarless of our current working directory.

Ensure you have the following files in a directory below the directory that holds your .proj file: RemoveQsubTempFiles.R, MakeQsubs.R, MakeJobs.R, and bashDirections.R (basically, cd to the directory that contains your R .proj file, make a new directory within that directory, then download the files to that new directory)

These files are required in order to run this next chunk as well as the chunks that involve running the QIIME2 commands through this R script. These files are available from Emily Giroux's github here: https://github.com/girouxem/oakwilt
```{r sourcing_my_functions, echo=TRUE, eval=TRUE, include=TRUE, tidy=FALSE}
library("rprojroot")
root        <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts     <- dir(root$find_file("R", path = root$find_file()))
scriptsl    <- paste(scriptsPath, scripts, sep = "/")
lapply(scriptsl, source)
```

Setting up working directories:
```{r}
sharedPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working"
analysis <- "oakwilt_redoITS"
sharedPathAn <- paste(sharedPath, analysis, sep = "/")
dir.create(sharedPathAn, showWarnings = TRUE, recursive = FALSE)
imageDirPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/GitHub_Repos/r_environments/oakwilt_redoITS"
dir.create("/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/GitHub_Repos/r_environments/oakwilt_redoITS", 
           showWarnings = TRUE, recursive = FALSE)
baseImage <- "oakwilt_redoITS.RData"
save.image(paste(imageDirPath, baseImage, sep = "/"))
```
Quick image load:
```{r}
imageDirPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/GitHub_Repos/r_environments/oakwilt_redoITS"
baseImage <- "oakwilt_redoITS.RData"
load(paste(imageDirPath, baseImage, sep = "/"))
```
### Step 1:       
Extract .fastq files from the compressed .tar.bz2 files.   
```{r}
library("data.table")
rawDataDirITSF <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/data/raw/Ion_Torrent/pirl_general/R_2021_01_05_00_59_57_user_S5-0143-150-OW_ITSF_Plate1InsectRpd_2021-01-04"
rawDataDirITS2R <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/data/raw/Ion_Torrent/pirl_general/R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04"
rawDataDir <- c(rawDataDirITSF, rawDataDirITS2R)
compressedFiles <- list.files(rawDataDir, pattern = "*.bz2", full.names = TRUE)
metadata <- as.data.table(cbind(compressedFiles))
metadata <- metadata[grep(".md5", metadata$compressedFiles, invert = TRUE) , ]
metadata$rawFileName <- basename(metadata$compressedFiles)
metadata$basename <- gsub(".tar.bz2", "", metadata$rawFileName)
rawDataWorkingPath <- paste(sharedPathAn, "rawData", sep = "/")
dir.create(rawDataWorkingPath, showWarnings = TRUE, recursive = FALSE)
metadata$rawWorkingPath <- paste(rawDataWorkingPath, metadata$basename, sep = "/")

for(i in 1:nrow(metadata)){
  cmd <- paste("mkdir -p ",  rawDataWorkingPath, " && tar -xvjf ", metadata$compressedFiles[i], 
                  " -C ", rawDataWorkingPath, sep = "")  
  system(cmd)
}

FwdDir <- paste(rawDataWorkingPath, "R_2021_01_05_00_59_57_user_S5-0143-150-OW_ITSF_Plate1InsectRpd_2021-01-04", sep = "/")
RevDir <- paste(rawDataWorkingPath, "R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04", sep = "/")
SeqDirs <- c(FwdDir, RevDir)
metadataITSF_files <- list.files(FwdDir, pattern = "ITSF", recursive = TRUE, full.names = TRUE)
metadataITSF <- as.data.table(cbind(metadataITSF_files))
metadataITSF$basename <- basename(metadataITSF$metadataITSF_files)
metadataITSF$barcode <- gsub(".*ITS1F_", "", metadataITSF$basename)
metadataITSF$barcode <- gsub(".fastq", "", metadataITSF$barcode)
metadataITSF$barcode <- gsub("b", "B", metadataITSF$barcode)

metadataITS2R_files <- list.files(RevDir, pattern = "ITS2rev", recursive = TRUE, full.names = TRUE)
metadataITS2R <- as.data.table(cbind(metadataITS2R_files))
metadataITS2R$basename <- basename(metadataITS2R$metadataITS2R_files)
metadataITS2R$barcode <- gsub(".*ITS2_A_", "", metadataITS2R$basename)
metadataITS2R$barcode <- gsub(".fastq", "", metadataITS2R$barcode)

# Join the metadata samples from the forward and reverse tables using the common barcode to join rows:
setkey(metadataITSF, barcode)
setkey(metadataITS2R, barcode)

metadataITS <- merge(metadataITSF, metadataITS2R, all.x = TRUE)
setnames(metadataITS, "basename.x", "fwdFastq")
setnames(metadataITS, "basename.y", "revFastq")
metadataITS <- na.omit(metadataITS)
```

Prepare file directories:
```{r}
# Make a directory for the trimmed data
trimmedData <- paste(sharedPathAn, "trimmed", sep = "/")
dir.create(trimmedData, showWarnings = TRUE, recursive = FALSE)

# Make a directory to hold the log files generated by itsxpress:
itsxpressLogs <- paste(sharedPathAn, "logs/itsxpress", sep = "/")
dir.create(itsxpressLogs, showWarnings = TRUE, recursive = TRUE)

# Make a directory to hold the output files generated by qiime2:
qiime2Dir <- paste(sharedPathAn, "qiime2", sep = "/")
dir.create(qiime2Dir, showWarnings = TRUE, recursive = FALSE)
```

### Step 2:       
Retrieve ITS1 part of the amplicons using ITSxpress (includes trimming regions and export)      
Run ITSxpress on the raw fastq reads:   
```{r}
trimmedDataITS1F <- paste(trimmedData, "trimmed_ITS1F", sep = "/")
dir.create(trimmedDataITS1F)

prefix <- "ITSxpress_ITS1F"
cmd <- paste("conda activate qiime2-2021.4 && itsxpress ",  
             " --fastq ", metadataITS$metadataITSF_files, 
             " --single_end ",
             " --outfile ", paste(trimmedDataITS1F, "/ITSF_trimmed.", metadataITS$barcode, ".fastq", sep = ""),
             " --region ITS1 --taxa Fungi --cluster_id 0.995 ",
             " --log ", paste(itsxpressLogs, "/ITSF_trimmed.", metadataITS$barcode, ".log", sep = ""),
             " && conda deactivate ",
             sep = "")  
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
**To remove the output files after you are done:**
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Before using ITSxpress on the ITS2R reads, use Marco's reverse compliment tool to reverse compliment them. This must be done because ITSxpress cannot recognize the reads in reverse-compliment orientation.

Tool available here: https://github.com/duceppemo/QIIME2_ITS/blob/master/fastq_rc.py
```{r}
ITS2revSeqsDir <- paste(rawDataWorkingPath, "R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04", sep = "/")
qiime2env <- paste("/home/CFIA-ACIA/gladishd/miniconda3/envs/qiime2-2021.4")
rcTool <- paste(qiime2env, "bin/fastq_rc.py", sep = "/")
rcToolOutputDir <- paste(rawDataWorkingPath, "RC_R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04", sep = "/")
dir.create(paste(rawDataWorkingPath, "RC_R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04", sep = "/"), showWarnings = TRUE, recursive    = FALSE)
```

```{r}
prefix <- "qiime2RC"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " && python3 ", rcTool, 
             " -i ", ITS2revSeqsDir,
             " -o ", rcToolOutputDir,
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)

metadataITS$metadataITS2R_RCfiles <- list.files(rcToolOutputDir, pattern = "ITS2", recursive = TRUE, full.names = TRUE)

## NOT WORKING, just run from command line for now by doing the following:
# cd /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/IonTorrent_Fungi_Barcoding_QIIME2
# python3 fastq_rc.py -i /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/rawData/R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04 -o /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/rawData/RC_R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04
```

```{r}
trimmedData2rev <- paste(trimmedData, "trimmed_ITS2R", sep = "/")
dir.create(trimmedData2rev)

prefix <- "ITSxpress_ITS2R"
cmd <- paste("conda activate qiime2-2021.4 && itsxpress ",  
             " --fastq ", metadataITS$metadataITS2R_RCfiles, 
             " --single_end ",
             " --outfile ", paste(trimmedData2rev, "/ITS2R_trimmed.", metadataITS$barcode, ".fastq", sep = ""),
             " --region ITS1 --taxa Fungi --cluster_id 0.995 ",
             " --log ", paste(itsxpressLogs, "/ITS2R_trimmed.", metadataITS$barcode, ".log", sep = ""),
             " && conda deactivate ",
             sep = "")  
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix) 
```

Add the path to the trimmed fastq files and a column to set unique sample names based on the filename/sequencing run and sample barcode number:
```{r}
metadataITS$trimmedPathITSF <- paste(trimmedDataITS1F, "/ITSF_trimmed.", metadataITSF$barcode, ".fastq", sep = "")
metadataITS$SampleID_ITSF <- paste("OW-ITSF-150-Plate1InsectRpd", metadataITS$barcode, sep = "-")

metadataITS$trimmedPathITS2R <- paste(trimmedData2rev, "/ITS2R_trimmed.", metadataITSF$barcode, ".fastq", sep = "")
metadataITS$SampleID_ITS2R <- paste("OW-ITS2R-151-Plate1InsectRpd", metadataITS$barcode, sep = "-")
```
## Importing the .fastq files into a QIIME2 usable format (.qza).

Create two separate manifest files that qiime2 will use to import our fastq data (ITS1F and ITS2R reads separately) and write it to two tsv files:

Manifest file: tells qiime2 the file path of the .fastq files (absolute-filepath) and what names they should have (sample-id).
```{r}
library("data.table")
manifestITS1F <- metadataITS[, .('sample-id' = SampleID_ITSF, 'absolute-filepath' = trimmedPathITSF)]
write.table(manifestITS1F, file = paste(sharedPathAn, "qiime2_import_manifestITS1F.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)

manifestITS2R <- metadataITS[, .('sample-id' = SampleID_ITS2R, 'absolute-filepath' = trimmedPathITS2R)]
write.table(manifestITS2R, file = paste(sharedPathAn, "qiime2_import_manifestITS2R.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```
Import fastq files as QIIME2 artifacts so that they may be used by the QIIME2 pipeline.
```{r}
prefix <- "qiimeImportITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             "qiime tools import ",
             " --type 'SampleData[SequencesWithQuality]' ",
             " --input-path ", paste(sharedPathAn, "qiime2_import_manifestITS1F.tsv", sep = "/"),
             " --output-path ", paste(qiime2Dir, "/demux-single-endITS1F.qza", sep = ""),
             " --input-format SingleEndFastqManifestPhred33V2 ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
prefix <- "qiimeImportITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             "qiime tools import ",
             " --type 'SampleData[SequencesWithQuality]' ",
             " --input-path ", paste(sharedPathAn, "qiime2_import_manifestITS2R.tsv", sep = "/"),
             " --output-path ", paste(qiime2Dir, "/demux-single-endITS2R.qza", sep = ""),
             " --input-format SingleEndFastqManifestPhred33V2 ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Create .qzv files to visualise the qiime2 fastq files imported:
```{r}
prefix <- "qiimeVisualiseImportITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime demux summarize ",
             " --i-data  ", paste(qiime2Dir, "/demux-single-endITS1F.qza", sep = ""),
             " --o-visualization ", paste(qiime2Dir, "/demux-single-endITS1F.qzv", sep = ""),
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
prefix <- "qiimeVisualiseImportITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime demux summarize ",
             " --i-data  ", paste(qiime2Dir, "/demux-single-endITS2R.qza", sep = ""),
             " --o-visualization ", paste(qiime2Dir, "/demux-single-endITS2R.qzv", sep = ""),
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
To view demux-single-end.qzv, open https://view.qiime2.org/ with your browser and drag the file into the window at the top of the page.     
    
**To remove the output files after you are done:**
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Denoise the sequences with dada2 within qiime2:    
- corrects sequencing errors    
- removes chimeras    
- clusters sequences at 100% similarity    
- outputs an amplicon sequence variant (ASV) table and the representative sequences
```{r}
prefix <- "qiimeDADA2deNoiseSingleITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime dada2 denoise-single ",
             " --i-demultiplexed-seqs ", paste(qiime2Dir, "/demux-single-endITS1F.qza", sep = ""),
             " --p-trim-left 0 ",
             " --p-trunc-len 0 ",
             " --o-representative-sequences ", paste(qiime2Dir, "/rep-seqs-dada2-ITS1F.qza", sep = ""),
             " --o-table ", paste(qiime2Dir, "/table-dada2-ITS1F.qza", sep = ""),
             " --o-denoising-stats ", paste(qiime2Dir, "/stats-dada2-ITS1F.qza", sep = ""),
             " --p-n-threads 20 ", 
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{R}
prefix <- "qiimeDADA2deNoiseSingleITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime dada2 denoise-single ",
             " --i-demultiplexed-seqs ", paste(qiime2Dir, "/demux-single-endITS2R.qza", sep = ""),
             " --p-trim-left 0 ",
             " --p-trunc-len 0 ",
             " --o-representative-sequences ", paste(qiime2Dir, "/rep-seqs-dada2-ITS2R.qza", sep = ""),
             " --o-table ", paste(qiime2Dir, "/table-dada2-ITS2R.qza", sep = ""),
             " --o-denoising-stats ", paste(qiime2Dir, "/stats-dada2-ITS2R.qza", sep = ""),
             " --p-n-threads 20 ", 
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Export the dada2 results:    
```{r}
# Export ASV (OTU-like table) table
prefix <- "qiimeExportITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " mkdir ", paste(sharedPathAn, "phyloseqITS1F", sep = "/"),
             " && qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/table-dada2-ITS1F.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseqITS1F", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Rename the output file "feature-table.biom" to "feature-table-ITS1F.biom"
```{r}
system("mv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS1F/feature-table.biom /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS1F/feature-table-ITS1F.biom")
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
prefix <- "qiimeExportITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " mkdir ", paste(sharedPathAn, "phyloseqITS2R", sep = "/"),
             " && qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/table-dada2-ITS2R.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseqITS2R", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Rename the output file "feature-table.biom" to "feature-table-ITS2R"
```{r}
system("mv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS2R/feature-table.biom /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS2R/feature-table-ITS2R.biom")
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
# Convert biom format to tsv format
prefix <- "qiimeBiomConvertITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " biom convert ",
             " -i ", paste(sharedPathAn, "phyloseqITS1F", "feature-table-ITS1F.biom", sep = "/"), 
             " -o ", paste(sharedPathAn, "phyloseqITS1F", "asv_tableITS1F.tsv", sep = "/"),
             " --to-tsv && cd ", paste(sharedPathAn, "phyloseqITS1F", sep = "/"),
             " && sed -i '1d' asv_tableITS1F.tsv && sed -i 's/#OTU ID//' asv_tableITS1F.tsv && cd .. ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
prefix <- "qiimeBiomConvertITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " biom convert ",
             " -i ", paste(sharedPathAn, "phyloseqITS2R", "feature-table-ITS2R.biom", sep = "/"), 
             " -o ", paste(sharedPathAn, "phyloseqITS2R", "asv_tableITS2R.tsv", sep = "/"),
             " --to-tsv && cd ", paste(sharedPathAn, "phyloseqITS2R", sep = "/"),
             " && sed -i '1d' asv_tableITS2R.tsv && sed -i 's/#OTU ID//' asv_tableITS2R.tsv && cd .. ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
# Export representative sequences
prefix <- "qiimeRepSeqsExportITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/rep-seqs-dada2-ITS1F.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseqITS1F", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Rename output file "dna-sequences.fasta" to "dna-sequences-ITS1F.fasta"
```{r}
system("mv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS1F/dna-sequences.fasta /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS1F/dna-sequences-ITS1F.fasta")
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
prefix <- "qiimeRepSeqsExportITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/rep-seqs-dada2-ITS2R.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseqITS2R", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Rename output file "dna-sequences.fasta" to "dna-sequences-ITS2R.fasta"
```{r}
system("mv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS2R/dna-sequences.fasta /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS2R/dna-sequences-ITS2R.fasta")
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Set up the qiime2 UNITE database using UNITE 2018-2019 that Marc-o used:
```{r}
uniteDBDir2019 <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/Databases/UNITE_2018-11-18_qiimeReleaseDB"
```

```{r}
# Import the UNITE reference sequences into QIIME2. Only have to put it in one of the phyloseq directories.
prefix <- "qiimeUniteImport"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime tools import ",
             " --type FeatureData[Sequence] ",
             " --input-path ", paste(uniteDBDir2019, "sh_refs_qiime_ver8_99_02.02.2019.fasta", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseqITS1F", "unite-ver8_99_02.02.2019.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

```{r}
# Import the taxonomy file.
prefix <- "qiimeTaxUniteImport"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime tools import ",
             " --type FeatureData[Taxonomy] ",
             " --input-path ", paste(uniteDBDir2019, "sh_taxonomy_qiime_ver8_99_02.02.2019.txt", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseqITS1F", "unite-ver8-taxonomy_99_02.02.2019.qza", sep = "/"),
             " --input-format HeaderlessTSVTaxonomyFormat ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

```{r}
# Train the classifier
prefix <- "qiimeTrainUnite"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime feature-classifier fit-classifier-naive-bayes ",
             " --i-reference-reads ", paste(sharedPathAn, "phyloseqITS1F", "unite-ver8_99_02.02.2019.qza", sep = "/"), 
             " --i-reference-taxonomy ", paste(sharedPathAn, "phyloseqITS1F", "unite-ver8-taxonomy_99_02.02.2019.qza", sep = "/"),
             " --o-classifier ", paste(sharedPathAn, "phyloseqITS1F", "unite-ver8-classifier_99_02.02.2019.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

Taxonomic classification using the ITS1F reads & the trained feature classifier
```{r}
prefix <- "qiimeClassifyFeatureITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime feature-classifier classify-sklearn ",
             " --i-classifier ", paste(sharedPathAn, "phyloseqITS1F", "unite-ver8-classifier_99_02.02.2019.qza", sep = "/"),
             " --i-reads ", paste(qiime2Dir, "rep-seqs-dada2-ITS1F.qza", sep = "/"),
             " --o-classification ", paste(qiime2Dir, "taxonomy-single-end-ITS1F.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

Taxonomic classification using the ITS2R reads & the trained feature classifier
```{r}
prefix <- "qiimeClassifyFeatureITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime feature-classifier classify-sklearn ",
             " --i-classifier ", paste(sharedPathAn, "phyloseqITS1F", "unite-ver8-classifier_99_02.02.2019.qza", sep = "/"),
             " --i-reads ", paste(qiime2Dir, "rep-seqs-dada2-ITS2R.qza", sep = "/"),
             " --o-classification ", paste(qiime2Dir, "taxonomy-single-end-ITS2R.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

Export ITS1F taxonomy .qza file into .tsv format
```{r}
prefix <- "qiimeFeatureExportITS1F"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "taxonomy-single-end-ITS1F.qza", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseqITS1F", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Rename to distinguish
```{r}
system("mv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS1F/taxonomy.tsv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS1F/taxonomy-ITS1F.tsv")
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Export ITS2R taxonomy .qza file into .tsv format
```{r}
prefix <- "qiimeFeatureExportITS2R"
cmd <- paste("conda activate qiime2-2021.4 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "taxonomy-single-end-ITS2R.qza", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseqITS2R", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Rename to distingusih
```{r}
system("mv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS2R/taxonomy.tsv /isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redoITS/phyloseqITS2R/taxonomy-ITS2R.tsv")
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Get the ASV table with taxonomy column at the end by combining the asv_tableITSxx.tsv and taxonomy-ITSxx.tsv tables together by the feature ID column
```{r}
library(data.table)
library(phyloseq)

asvTbl_ITS1F <- fread(paste(sharedPathAn, "phyloseqITS1F", "asv_tableITS1F.tsv", sep = "/"))
taxTbl_ITS1F <- fread(paste(sharedPathAn, "phyloseqITS1F", "taxonomy-ITS1F.tsv", sep = "/"))

asvTbl_ITS2R <- fread(paste(sharedPathAn, "phyloseqITS2R", "asv_tableITS2R.tsv", sep = "/"))
taxTbl_ITS2R <- fread(paste(sharedPathAn, "phyloseqITS2R", "taxonomy-ITS2R.tsv", sep = "/"))

# Open the taxonomy and change the header. When you open it, youâ€™ll see the header looks like this: 
# Feature ID	Taxon	Confidence
# where the spaces are tabs. You need to change it to this:
# asv-id	taxonomy	Confidence

setnames(taxTbl_ITS1F, "Feature ID", "asv-id")
setnames(taxTbl_ITS1F, "Taxon", "taxonomy")

setnames(taxTbl_ITS2R, "Feature ID", "asv-id")
setnames(taxTbl_ITS2R, "Taxon", "taxonomy")


setnames(asvTbl_ITS1F, "V1", "asv-id")

setnames(asvTbl_ITS2R, "V1", "asv-id")


setkey(asvTbl_ITS1F, "asv-id")
setkey(taxTbl_ITS1F, "asv-id")

setkey(asvTbl_ITS2R, "asv-id")
setkey(taxTbl_ITS2R, "asv-id")

asvTaxTbl_ITS1F <- merge(asvTbl_ITS1F, taxTbl_ITS1F)
asvTaxTbl_ITS1F$Confidence <- NULL
asvTaxTbl_ITS1F$taxonomy <- gsub(";", "; ", asvTaxTbl_ITS1F$taxonomy)
write.table(asvTaxTbl_ITS1F, file = paste(sharedPathAn, "asvTax_tableITS1F.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)

asvTaxTbl_ITS2R <- merge(asvTbl_ITS2R, taxTbl_ITS2R)
asvTaxTbl_ITS2R$Confidence <- NULL
asvTaxTbl_ITS2R$taxonomy <- gsub(";", "; ", asvTaxTbl_ITS2R$taxonomy)
write.table(asvTaxTbl_ITS2R, file = paste(sharedPathAn, "asvTax_tableITS2R.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```

```{r}
library("RAM")
dir.create(paste(sharedPathAn, "taxFill", sep = "/"), showWarnings = TRUE, recursive    = FALSE)
taxFillPath <- paste(sharedPathAn, "taxFill", sep = "/")

tempITS1F <- read.table(paste(sharedPathAn, "asvTax_tableITS1F.tsv", sep = "/"), sep = "\t", header = TRUE, dec = ".", 
                   comment.char = "", quote = "", stringsAsFactors = TRUE,
                   as.is = TRUE, colClasses=c("taxonomy"="character"))

row.names(tempITS1F) <- tempITS1F$asv.id
tempITS1F$asv.id <- NULL

tempITS1F <- tax.fill(tempITS1F, downstream=TRUE)

write.table(tempITS1F, file=paste(taxFillPath, "ITS1F.table.taxfill.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote = FALSE)
```
```{r}
tempITS2R <- read.table(paste(sharedPathAn, "asvTax_tableITS2R.tsv", sep = "/"), sep = "\t", header = TRUE, dec = ".", 
                   comment.char = "", quote = "", stringsAsFactors = TRUE,
                   as.is = TRUE, colClasses=c("taxonomy"="character"))

row.names(tempITS2R) <- tempITS2R$asv.id
tempITS2R$asv.id <- NULL

tempITS2R <- tax.fill(tempITS2R, downstream=TRUE)

write.table(tempITS2R, file=paste(taxFillPath, "ITS2R.table.taxfill.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote = FALSE)
```
Merge the two taxfilled tables into one excel file (one file that has two sheets).

Might also be able to use a table merging technique as an alternative to this to maintain consistency of use of the .tsv format. Could be worth looking into.
```{r}
install.packages("openxlsx")
library("openxlsx")

taxfillJoined <- createWorkbook()
addWorksheet(taxfillJoined, "ITS1F.table.taxfill")
addWorksheet(taxfillJoined, "ITS2R.table.taxfill")

writeData(taxfillJoined, "ITS1F.table.taxfill", tempITS1F)
writeData(taxfillJoined, "ITS2R.table.taxfill", tempITS2R)

saveWorkbook(taxfillJoined, file = paste(taxFillPath, "ASVJoinedTaxfillTable.xlsx", sep = "/"), overwrite = TRUE)
```


Create diversity directory and path. List your datasets for diversity indexes. Adds a bunch of columns in a new diversity metadata file.
Indices obtained (in this specific order) are: Spec Number,  Simpson data,  Inv simpson data,	Shannon data,	Simpson eveness,	Shannon eveness,	Simpson true diversity,	shannon true diversity,	chao,	ACE.
```{r}
dir.create(paste(sharedPathAn, "diversity", sep = "/"), showWarnings = TRUE, recursive = FALSE)
diversityPath <- paste(sharedPathAn, "diversity", sep = "/")

metaTempITS1F <- metadataITSF
rownames(metaTemp) <- colnames(tempITS1F)[-ncol(tempITS1F)]
temp2_ITS1F <- OTU.diversity(list(data=tempITS1F), metaTempITS1F)

write.table(temp2_ITS1F, file=paste(diversityPath, "OW-ITS1F-S5-0143-150.meta.div.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote=FALSE)


metaTempITS2R <- metadataITS2R
rownames(metaTempITS2R) <- colnames(tempITS2R)[-ncol(tempITS2R)]
temp2_ITS2R <- OTU.diversity(list(data=tempITS2R), metaTempITS2R)

write.table(temp2_ITS1F, file=paste(diversityPath, "OW-ITS2R-S5-0143-151.meta.div.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote=FALSE)
```
Combine the two diversity index tables into one excel file composed of 2 sheets
```{r}
diversityJoined <- createWorkbook()
addWorksheet(diversityJoined, "OW-ITS1F-S5-0143-150.meta.div")
addWorksheet(diversityJoined, "OW-ITS2R-S5-0143-151.meta.div")

writeData(diversityJoined, "OW-ITS1F-S5-0143-150.meta.div", temp2_ITS1F)
writeData(diversityJoined, "OW-ITS2R-S5-0143-151.meta.div", temp2_ITS2R)

saveWorkbook(diversityJoined, file = paste(diversityPath, "DivJoinedTaxfillTable.xlsx", sep = "/"), overwrite = TRUE)
```

Save image:
```{r}
save.image(paste(imageDirPath, baseImage, sep = ""))
```
FUNGuildR: https://github.com/brendanf/FUNGuildR/

Obtain guild information and to append it to the asv table.
```{r}
FUNGuildDir <- paste(sharedPathAn, "FUNGuild", sep = "/")
dir.create(paste(sharedPathAn, "FUNGuild", sep = "/"),
             showWarnings = TRUE,
             recursive    = FALSE)

library("usethis")
library("devtools")
library(knitr)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
knit_print.data.frame = function(x, ...) {
    res = paste(c("", "", knitr::kable(x, output = FALSE)), collapse = "\n")
    asis_output(res)
}
# register the method
registerS3method("knit_print", "data.frame", knit_print.data.frame)

devtools::install_github("brendanf/FUNGuildR")
library("FUNGuildR")
```

```{r}
OW_OTU_ITS1F <- tempITS1F
OW_guildsITS1F <- funguild_assign(otu_table = OW_OTU_ITS1F, db = get_funguild_db(), tax_col = "taxonomy")

write.table(OW_guildsITS1F, file=paste(FUNGuildDir, "asvtbl_OW-ITS1F-S5-0143-150.guilds.tsv", sep = "/"),
              append    = FALSE,
              sep       = "\t",
              row.names = FALSE,
              quote=FALSE)



OW_OTU_ITS2R <- tempITS2R
OW_guildsITS2R <- funguild_assign(otu_table = OW_OTU_ITS2R, db = get_funguild_db(), tax_col = "taxonomy")

write.table(OW_guildsITS1F, file=paste(FUNGuildDir, "asvtbl_OW-ITS2R-S5-0143-151.guilds.tsv", sep = "/"),
              append    = FALSE,
              sep       = "\t",
              row.names = FALSE,
              quote=FALSE)
```
Combine the two ASV tables containing guild information together
```{r}
guildJoined <- createWorkbook()
addWorksheet(guildJoined, "OW-ITS1F-S5-0143-150.guilds")
addWorksheet(guildJoined, "OW-ITS2R-S5-0143-151.guilds")

writeData(guildJoined, "OW-ITS1F-S5-0143-150.guilds", OW_guildsITS1F)
writeData(guildJoined, "OW-ITS2R-S5-0143-151.guilds", OW_guildsITS2R)

saveWorkbook(guildJoined, file = paste(FUNGuildDir, "guildsJoinedTaxfillTable.xlsx", sep = "/"), overwrite = TRUE)
```
