---
title: "OakWilt_RedoITS1F"
author: "Daniel Gladish"
date: "1/19/2021"
output:
  pdf_document: default
  html_document: default
urlcolor: blue
header-includes: \usepackage{xcolor}
---

Set .libPaths() to the directory that houses the downloaded R packages.
```{r}
.libPaths( c( "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/R/x86_64-pc-linux-gnu-library/3.5" , .libPaths() ) )

```


```{r, global_options, eval=TRUE, echo=FALSE}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff = 80), tidy = TRUE, fig.align = 'center',
               cache = FALSE, collapse = TRUE, echo = FALSE, eval = FALSE, include = FALSE,
               message = FALSE, quietly = TRUE, results = 'hide', warn.conflicts = FALSE, 
               warning = FALSE)
```

**Using package `BiocManager` to install required packages:**
```{r, biocInstall, eval=TRUE, echo=TRUE, include=TRUE}
#Installing required packages
r <- getOption("repos")
r["CRAN"] <- "http://cran.us.r-project.org"
options(repos = r)

if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()

library("BiocManager")
.cran_packages <- c("data.table", "kableExtra", "knitr", "rprojroot")
.bioc_packages <- c("BiocStyle", "Biostrings", "dada2", "RAM")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  BiocManager::install(.bioc_packages[!.inst], ask = FALSE)
}
```
   
**Load packages into session, and print package versions:**
```{r, showBiocPackages, echo=TRUE, eval=TRUE, include=TRUE, results='hold'}
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
```
**Source our custom R scripts:**    
For this we will use the rprojroot package to set the directory structures. This will help us when finding our files to source functions. We specify ours is an RStudio project. The root object contains a function that will help us locate our package R files regarless of our current working directory.

Ensure you have the following files in a directory below the directory that holds your .proj file: RemoveQsubTempFiles.R, MakeQsubs.R, MakeJobs.R, and bashDirections.R (basically, cd to the directory that contains your R .proj file, make a new directory within that directory, then download the files to that new directory). Ensure that the name of the directory is "R".

These files are required in order to run this next chunk as well as the chunks that involve running the QIIME2 commands through this R script. These files are available from Emily Giroux's github oakwilt repository here: https://github.com/girouxem/oakwilt
```{r sourcing_my_functions, echo=TRUE, eval=TRUE, include=TRUE, tidy=FALSE}
library("rprojroot")
root        <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts     <- dir(root$find_file("R", path = root$find_file()))
scriptsl    <- paste(scriptsPath, scripts, sep = "/")
lapply(scriptsl, source)
```

Set up working directories
```{r}
sharedPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working"
analysis <- "oakwilt_redo"
sharedPathAn <- paste(sharedPath, analysis, sep = "/")
dir.create(sharedPathAn, showWarnings = TRUE, recursive = FALSE)
imageDirPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/GitHub_Repos/r_environments/oakwilt_redo/"
dir.create("/isilon/cfia-ottawa-fallowfield/users/girouxeml//gladishd/GitHub_Repos/r_environments/oakwilt_redo", 
           showWarnings = TRUE, recursive = FALSE)
baseImage <- "oakwilt_redo.RData"
save.image(paste(imageDirPath, baseImage, sep = ""))
```
Quick image load
```{r}
imageDirPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/GitHub_Repos/r_environments/oakwilt_redo/"
baseImage <- "oakwilt_redo.RData"
load(paste(imageDirPath, baseImage, sep = ""))
```

Extract .fastq files from the compressed .tar.bz2 files.
```{r}
library("data.table")
rawDataDirITSF <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/data/raw/Ion_Torrent/pirl_general/R_2021_01_05_00_59_57_user_S5-0143-150-OW_ITSF_Plate1InsectRpd_2021-01-04"
rawDataDirITS2R <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/data/raw/Ion_Torrent/pirl_general/R_2021_01_05_06_44_57_user_S5-0143-151-OW_ITS2rev_Plate1Insect_repeated_2021-01-04"
rawDataDir <- c(rawDataDirITSF, rawDataDirITS2R)
compressedFiles <- list.files(rawDataDir, pattern = "*.bz2", full.names = TRUE)
metadata <- as.data.table(cbind(compressedFiles))
metadata <- metadata[grep(".md5", metadata$compressedFiles, invert = TRUE) , ]
metadata$rawFileName <- basename(metadata$compressedFiles)
metadata$basename <- gsub(".tar.bz2", "", metadata$rawFileName)
rawDataWorkingPath <- paste(sharedPathAn, "rawData", sep = "/")
dir.create(rawDataWorkingPath, showWarnings = TRUE, recursive = FALSE)
metadata$rawWorkingPath <- paste(rawDataWorkingPath, metadata$basename, sep = "/")

for(i in 1:nrow(metadata)){
  cmd <- paste("mkdir -p ",  rawDataWorkingPath, " && tar -xvjf ", metadata$compressedFiles[i], 
                  " -C ", rawDataWorkingPath, sep = "")  
  system(cmd)
}

metadataITSF_files <- list.files(rawDataWorkingPath, pattern = "ITSF", recursive = TRUE, full.names = TRUE)
metadataITSF <- as.data.table(cbind(metadataITSF_files))
metadataITSF$basename <- basename(metadataITSF$metadataITSF_files)
metadataITSF$barcode <- gsub(".*ITS1F_", "", metadataITSF$basename)
metadataITSF$barcode <- gsub(".fastq", "", metadataITSF$barcode)
metadataITSF$barcode <- gsub("b", "B", metadataITSF$barcode)

metadataITS2R_files <- list.files(rawDataWorkingPath, pattern = "ITS2rev", recursive = TRUE, full.names = TRUE)
metadataITS2R <- as.data.table(cbind(metadataITS2R_files))
metadataITS2R$basename <- basename(metadataITS2R$metadataITS2R_files)
metadataITS2R$barcode <- gsub(".*ITS2_A_", "", metadataITS2R$basename)
metadataITS2R$barcode <- gsub(".fastq", "", metadataITS2R$barcode)

# Join the metadata samples from the forward and reverse tables using the common barcode to join rows:
setkey(metadataITSF, barcode)
setkey(metadataITS2R, barcode)

metadataITS <- merge(metadataITSF, metadataITS2R, all.x = TRUE)
setnames(metadataITS, "basename.x", "fwdFastq")
setnames(metadataITS, "basename.y", "revFastq")
metadataITS <- na.omit(metadataITS)
```

Prepare directories that will house the ITSxpress and QIIME2 output files
```{r}
# Make a directory for the trimmed data
trimmedData <- paste(sharedPathAn, "trimmed", sep = "/")
dir.create(trimmedData, showWarnings = TRUE, recursive = FALSE)

# Make a directory to hold the log files generated by itsxpress:
itsxpressLogs <- paste(sharedPathAn, "logs/itsxpress", sep = "/")
dir.create(itsxpressLogs, showWarnings = TRUE, recursive = TRUE)

qiime2Dir <- paste(sharedPathAn, "qiime2", sep = "/")
dir.create(qiime2Dir, showWarnings = TRUE, recursive = FALSE)
```

### Step 2:       
Retrieve ITS1 part of the amplicons using ITSxpress (includes trimming regions and export)      
Run ITSxpress on the raw fastq reads:   
     
**Note:** For the ITS2 region, itsxpress does not recognise the ITS2 amplicon regions and nothing is returned - no OTU table for ITS2 can be generated. We are currently looking into finding a solution so that the ITS2R reads can be used as well.
```{r}
prefix <- "ITSxpress_ITSF"
cmd <- paste("conda activate qiime2-2021.2 && itsxpress ",  
             " --fastq ", metadataITS$metadataITSF_files, 
             " --single_end ",
             " --outfile ", paste(trimmedData, "/ITSF_trimmed.", metadataITS$barcode, ".fastq", sep = ""),
             " --region ITS1 --taxa Fungi --cluster_id 0.995 ",
             " --log ", paste(itsxpressLogs, "/ITSF_trimmed.", metadataITS$barcode, ".log", sep = ""),
             " && conda deactivate ",
             sep = "")  
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
**To remove the output files after you are done:**
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Add the path to the trimmed fastq files and a column to set unique sample names based on the filename/sequencing run and sample barcode number
```{r}
metadataITSF$trimmedPath <- paste(trimmedData, "/ITSF_trimmed.", metadataITSF$barcode, ".fastq", sep = "")
metadataITSF$SampleID <- paste("OW-ITSF-2021-Plate1InsectRpd", metadataITSF$barcode, sep = "-")
```

Create a manifest file that qiime2 will use to import our fastq data and write it to a tsv file
```{r}
library("data.table")
manifest <- metadataITSF[, .('sample-id' = SampleID, 'absolute-filepath' = trimmedPath)]

write.table(manifest, file = paste(sharedPathAn, "qiime2_import_manifest.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```

Import fastq files
```{r}
prefix <- "qiimeImport"
cmd <- paste("conda activate qiime2-2021.2 && ",
             "qiime tools import ",
             " --type 'SampleData[SequencesWithQuality]' ",
             " --input-path ", paste(sharedPathAn, "qiime2_import_manifest.tsv", sep = "/"),
             " --output-path ", paste(qiime2Dir, "/demux-single-end.qza", sep = ""),
             " --input-format SingleEndFastqManifestPhred33V2 ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Create a file to visualise the qiime2 fastq files imported
```{r}
prefix <- "qiimeVisualiseImport"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime demux summarize ",
             " --i-data  ", paste(qiime2Dir, "/demux-single-end.qza", sep = ""),
             " --o-visualization ", paste(qiime2Dir, "/demux-single-end.qzv", sep = ""),
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
To view demux-single-end.qzv, open https://view.qiime2.org/ with your browser and drag the file into the window at the top of the page.     
    
**To remove the output files after you are done:**
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Denoise the sequences with dada2 within qiime2:    
- corrects sequencing errors    
- removes chimeras    
- clusters sequences at 100% similarity    
- outputs an asv table and the representative sequences
```{r}
prefix <- "qiimeDADA2deNoiseSingle"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime dada2 denoise-single ",
             " --i-demultiplexed-seqs ", paste(qiime2Dir, "/demux-single-end.qza", sep = ""),
             " --p-trim-left 0 ",
             " --p-trunc-len 0 ",
             " --o-representative-sequences ", paste(qiime2Dir, "/rep-seqs-dada2.qza", sep = ""),
             " --o-table ", paste(qiime2Dir, "/table-dada2.qza", sep = ""),
             " --o-denoising-stats ", paste(qiime2Dir, "/stats-dada2.qza", sep = ""),
             " --p-n-threads 20 ", 
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Export the dada2 results
```{r}
# Export ASV (OTU-like table) table
prefix <- "qiimeExport"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " mkdir ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/table-dada2.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
# Convert biom format to tsv format
prefix <- "qiimeBiomConvert"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " biom convert ",
             " -i ", paste(sharedPathAn, "phyloseq", "feature-table.biom", sep = "/"), 
             " -o ", paste(sharedPathAn, "phyloseq", "otu_table.tsv", sep = "/"),
             " --to-tsv && cd ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && sed -i '1d' otu_table.tsv && sed -i 's/#OTU ID//' otu_table.tsv && cd .. ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
# Export representative sequences
prefix <- "qiimeRepSeqsExport"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/rep-seqs-dada2.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```


Set up the qiime2 UNITE database
```{r}
uniteDBDir2019 <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/Databases/UNITE_2018-11-18_qiimeReleaseDB"
```

```{r}
# Import the UNITE reference sequences into QIIME2.
prefix <- "qiimeUniteImport"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime tools import ",
             " --type FeatureData[Sequence] ",
             " --input-path ", paste(uniteDBDir2019, "sh_refs_qiime_ver8_99_02.02.2019.fasta", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseq", "unite-ver8_99_02.02.2019.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```


```{r}
# Import the taxonomy file.
prefix <- "qiimeTaxUniteImport"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime tools import ",
             " --type FeatureData[Taxonomy] ",
             " --input-path ", paste(uniteDBDir2019, "sh_taxonomy_qiime_ver8_99_02.02.2019.txt", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseq", "unite-ver8-taxonomy_99_02.02.2019.qza", sep = "/"),
             " --input-format HeaderlessTSVTaxonomyFormat ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

```{r}
# Train the classifier
prefix <- "qiimeTrainUnite"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime feature-classifier fit-classifier-naive-bayes ",
             " --i-reference-reads ", paste(sharedPathAn, "phyloseq", "unite-ver8_99_02.02.2019.qza", sep = "/"), 
             " --i-reference-taxonomy ", paste(sharedPathAn, "phyloseq", "unite-ver8-taxonomy_99_02.02.2019.qza", sep = "/"),
             " --o-classifier ", paste(sharedPathAn, "phyloseq", "unite-ver8-classifier_99_02.02.2019.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

```{r}
prefix <- "qiimeClassifyFeature"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime feature-classifier classify-sklearn ",
             " --i-classifier ", paste(sharedPathAn, "phyloseq", "unite-ver8-classifier_99_02.02.2019.qza", sep = "/"),
             " --i-reads ", paste(qiime2Dir, "rep-seqs-dada2.qza", sep = "/"),
             " --o-classification ", paste(qiime2Dir, "taxonomy-single-end.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

```{r}
prefix <- "qiimeFeatureExport"
cmd <- paste("conda activate qiime2-2021.2 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "taxonomy-single-end.qza", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

Obtain the ASV/OTU table
```{r}
list.files(path = paste(sharedPathAn, "phyloseq", sep = "/"))
library(data.table)
library(phyloseq)
otuTbl <- fread(paste(sharedPathAn, "phyloseq", "otu_table.tsv", sep = "/"))
taxTbl <- fread(paste(sharedPathAn, "phyloseq", "taxonomy.tsv", sep = "/"))


# Open the taxonomy and change the header. When you open it, you’ll see the header looks like this: 
# Feature ID	Taxon	Confidence
# where the spaces are tabs. You need to change it to this:
# otu-id	taxonomy	Confidence

setnames(taxTbl, "Feature ID", "otu-id")
setnames(taxTbl, "Taxon", "taxonomy")

setnames(otuTbl, "V1", "otu-id")

setkey(otuTbl, "otu-id")
setkey(taxTbl, "otu-id")

otuTaxTbl <- merge(otuTbl, taxTbl)
otuTaxTbl$Confidence <- NULL

write.table(otuTaxTbl, file = paste(sharedPathAn, "otuTax_table.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```

```{r}
library("RAM")
dir.create(paste(sharedPathAn, "taxFill", sep = "/"), showWarnings = TRUE, recursive    = FALSE)
taxFillPath <- paste(sharedPathAn, "taxFill", sep = "/")

temp <- read.table(paste(sharedPathAn, "otuTax_table.tsv", sep = "/"), sep = "\t", header = TRUE, dec = ".", 
                   comment.char = "", quote = "", stringsAsFactors = TRUE,
                   as.is = TRUE, colClasses=c("taxonomy"="character"))

row.names(temp) <- temp$otu.id
temp$otu.id <- NULL


## for later when an OTU that has not been subjected to the tax.fill() function is required. Note that doing this write.table() function removes the row names, removing the otu IDs, which is ok for the downstream analysis. ##

notaxfill <- temp

write.table(notaxfill, file = paste(sharedPathAn, "otuTax_tableOW_redo_NoTaxfill.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)

## continue with script ##


temp <- tax.fill(temp, downstream=TRUE)

write.table(temp, file=paste(taxFillPath, "ITS1F.table.taxfill.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote = FALSE)
## ^^ note that the row.names argument = FALSE, therefore row names will be removed.
```


Create diversity directory and path. List your datasets for diversity indexes. Adds a bunch of columns in a new diversity metadata file.

Indices obtained (in this specific order) are: Spec Number,  Simpson data,  Inv simpson data,	Shannon data,	Simpson eveness,	Shannon eveness,	Simpson true diversity,	shannon true diversity,	chao,	ACE.
```{r}
dir.create(paste(sharedPathAn, "diversity", sep = "/"), showWarnings = TRUE, recursive = FALSE)
diversityPath <- paste(sharedPathAn, "diversity", sep = "/")

metaTemp <- metadataITSF
rownames(metaTemp) <- colnames(temp)[-ncol(temp)]
temp2 <- OTU.diversity(list(data=temp), metaTemp)

write.table(temp2, file=paste(diversityPath, "owRedo2021_ITS1FPlate1.meta.div.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote=FALSE)
```

Save image
```{r}
save.image(paste(imageDirPath, baseImage, sep = ""))
```
FUNGuildR

Obtain guild information and to append it to the OTU table. Code obtained from https://github.com/brendanf/FUNGuildR/. 
```{r}
FUNGuildDir <- paste(sharedPathAn, "FUNGuild", sep = "/")
dir.create(paste(sharedPathAn, "FUNGuild", sep = "/"),
             showWarnings = TRUE,
             recursive    = FALSE)

library("devtools")
library(knitr)
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
knit_print.data.frame = function(x, ...) {
    res = paste(c("", "", knitr::kable(x, output = FALSE)), collapse = "\n")
    asis_output(res)
}
# register the method
registerS3method("knit_print", "data.frame", knit_print.data.frame)

devtools::install_github("brendanf/FUNGuildR")
library("FUNGuildR")
```

```{r}
OW_OTU <- notaxfill
OW_guilds <- funguild_assign(otu_table = OW_OTU, db = get_funguild_db(), tax_col = "taxonomy")

write.table(OW_guilds, file=paste(FUNGuildDir, "otuTax_tableOW_redo.guilds.tsv", sep = "/"),
              append    = FALSE,
              sep       = "\t",
              row.names = FALSE,
              quote=FALSE)
```

FIGURES
```{r}
library("RAM")

sharedPathAn <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redo"

taxFillPath <- paste(sharedPathAn, "taxFill", sep = "/")

tftemp <- read.table(paste(taxFillPath, "ITS1F.table.taxfill.tsv", sep = "/"),
                      sep = "\t", header = TRUE, dec = ".", 
                      comment.char = "", 
                      quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE, colClasses=c("taxonomy"="character"))

tftemp <- tax.fill(temptest, downstream=TRUE)
# ^^ If you look at the "taxonomy" column after running this line, you will see that the tax.fill function removes spaces between the taxonomic levels, and causes repeats of names. While the tax.fill() function removes incertae sedis, this causes errors when trying to make the figures. Therefore, it is recommended that the pre-tax.fill() version of the OTU table be used to create figures whose associated functions require OTU table data as input (i.e. the abundance plots and venn diagrams).


```

```{r}
metatemp <-  read.table(paste(sharedPathAn, "OWPlate1_ITS1F.meta.tsv", sep = "/"),
                      sep = "\t", header = TRUE, dec = ".", comment.char = "", quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE)

metatemp$row.names <- NULL

rownames(metatemp) <- colnames(tftemp)[-ncol(tftemp)]

```

Calculation of diversity indices
```{r}
dir.create(paste(sharedPathAn, "diversity", sep = "/"),
           showWarnings = TRUE,
           recursive  = FALSE)
diversityPath <- paste(sharedPathAn, "diversity", sep = "/")

metatempdiv <- OTU.diversity(list(data=tftemp), metatemp)

write.table(metatempdiv, file=paste(sharedPathAn, "OW2021redo_ITS1F.meta.div.tsv", sep = "/"),
              append    = FALSE,
              sep       = "\t",
              row.names = FALSE,
              quote=FALSE)
```

Mean & whisker plots
```{r}

dir.create(paste(diversityPath, "diversityPlot", sep = "/"),
           showWarnings = TRUE,
           recursive    = FALSE)
diversityPlotPath <- paste(diversityPath, "diversityPlot", sep = "/")


#### Shannon True Div ###

## TrapLocation ##

group.diversity(list(data=tftemp), metatempdiv, factors = c("TrapLocation"), indices = c("shan_trudiv"), diversity.info=TRUE, compare="NULL", ylab="Shannon True Div", facet.x.cex = "10", facet.y.cex = "10")


# Run the group.diversity() function in the console. The figure will appear in the bottom right. Export and save as image (.png) to directory "diversity plot" with appropriate dimensions. Dimensions 800 x 700 seems to work well.



## CollectedLocation ##

group.diversity(list(data=tftemp), metatempdiv, factors = c("CollectedLocation"), indices = c("shan_trudiv"), diversity.info=TRUE, compare="NULL", ylab="Shannon True Div", facet.x.cex = "10", facet.y.cex = "10")


### Simpson True Diversity ###

## TrapLocation ##

group.diversity(list(data=tftemp), metatempdiv, factors = c("TrapLocation"), indices = c("sim_trudiv"), diversity.info=TRUE, compare="NULL", ylab="Simpson True Div", facet.x.cex = "10", facet.y.cex = "10")


## CollectedLocation ##

group.diversity(list(data=tftemp), metatempdiv, factors = c("CollectedLocation"), indices = c("sim_trudiv"), diversity.info=TRUE, compare="NULL", ylab="Simpson True Div", facet.x.cex = "10", facet.y.cex = "10")


### Simpson Evenness ###

## TrapLocation ##

group.diversity(list(data=tftemp), metatempdiv, factors = c("TrapLocation"), indices = c("sim_even"), diversity.info=TRUE, compare="NULL", ylab="Simpson Evenness", facet.x.cex = "10", facet.y.cex = "10")


## CollectedLocation ##

group.diversity(list(data=tftemp), metatempdiv, factors = c("CollectedLocation"), indices = c("sim_even"), diversity.info=TRUE, compare="NULL", ylab="Simpson Evenness", facet.x.cex = "10", facet.y.cex = "10")
```

Generate abundance bar graphs
```{r}
visdir <- paste(sharedPathAn, "visualizations", sep = "/")
dir.create(paste(visdir, "groupAbundanceBarPlots", sep = "/"),
             showWarnings = TRUE,
             recursive    = FALSE)

abundNbBarPlotPath <- paste(visdir, "groupAbundanceBarPlots", sep = "/")


# Using the .tsv files that had been subjected to tax.fill() results in names that are too long. This can interfere with the labels of the abundance plots. Using table "notaxfill" should work but it needs spaces between the taxonomic levels in the "taxonomy" column. Write some code to give spaces in notaxfill to solve these issues.

notaxfill$taxonomy <- gsub(";", "; ", notaxfill$taxonomy)


#### TrapLocation ###

## Species ##

group.abundance.meta(data=(list(data=notaxfill)), rank="s", top = 30, count =  TRUE, drop.unclassified = FALSE, cex.x = 6.7, main = "Top 30 counts of taxonomic groups at the species level", meta = metatemp, meta.factor = c("TrapLocation"))

# In the Plot menu, click export then save as image. Save as .png with dimensions width = 3500, height = 2000. Do this saving procedure after each time group.abundance.meta() is run. Save to created directory "groupAbundanceBarPlots".


## Genus ##

group.abundance.meta(data=(list(data=notaxfill)), rank="g", top = 30, count =  TRUE, drop.unclassified = FALSE, cex.x = 6.7, main = "Top 30 counts of taxonomic groups at the genus level", meta = metatemp, meta.factor = c("TrapLocation"))


#### Collected Location ###


## Species ##

group.abundance.meta(data=(list(data=notaxfill)), rank="s", top = 30, count =  TRUE, drop.unclassified = FALSE, cex.x = 7.5, main = "Top 30 counts of taxonomic groups at the species level", meta = metatemp, meta.factor = c("CollectedLocation"))


## Genus ##

group.abundance.meta(data=(list(data=notaxfill)), rank="s", top = 30, count =  TRUE, drop.unclassified = FALSE, cex.x = 7.5, main = "Top 30 counts of taxonomic groups at the species level", meta = metatemp, meta.factor = c("CollectedLocation"))

```


Venn Diagrams
```{r}

# When doing this chunk, run the group.venn() steps by copy pasting the code into the console. Click export and save the file as a .png in directory "visualizations/vennDiagrams" with appropriate dimensions. Dimensions that worked well for me are width = 2500 and height = 1500.

## TrapLocation - Species

coreTaxaITSTrapLoca_S <- core.Taxa(data = list(data = notaxfill), is.OTU = TRUE, meta = metatemp, rank = "s", drop.unclassified = FALSE, meta.factor = "TrapLocation", percent = 0)

Trap1Taxa_S <- coreTaxaITSTrapLoca_S$data$Trap1$taxa
Trap2Taxa_S <- coreTaxaITSTrapLoca_S$data$Trap2$taxa
Trap3Taxa_S <- coreTaxaITSTrapLoca_S$data$Trap3$taxa
CourTaxa_S <- coreTaxaITSTrapLoca_S$data$Cour$taxa

TrapLocaVector_S <- list(Trap1 = Trap1Taxa_S, Trap2 = Trap2Taxa_S, Trap3 = Trap3Taxa_S, Cour = CourTaxa_S)

group.venn(vectors = TrapLocaVector_S, cat.cex = 3, cex = 3, label = FALSE, lab.cex = 1, fill = c("red", "blue", "green", "pink"), lab.col = "black")


## TrapLocation - Genus

coreTaxaITSTrapLoca_G <- core.Taxa(data = list(data = notaxfill), is.OTU = TRUE, meta = metatemp, rank = "g", drop.unclassified = FALSE, meta.factor = "TrapLocation", percent = 0)

Trap1Taxa_G <- coreTaxaITSTrapLoca_G$data$Trap1$taxa
Trap2Taxa_G <- coreTaxaITSTrapLoca_G$data$Trap2$taxa
Trap3Taxa_G <- coreTaxaITSTrapLoca_G$data$Trap3$taxa
CourTaxa_G <- coreTaxaITSTrapLoca_G$data$Cour$taxa

TrapLocaVector_G <- list(Trap1 = Trap1Taxa_G, Trap2 = Trap2Taxa_G, Trap3 = Trap3Taxa_G, Cour = CourTaxa_G)

group.venn(vectors = TrapLocaVector_G, cat.cex = 3, cex = 3, label = FALSE, lab.cex = 1, fill = c("red", "blue", "green", "pink"), lab.col = "black")

```

Figure out abundance rank of Ceratocystis fagacearum in each sample that shows presence of this pathogen.
```{r}
install.packages("tidyverse")
library("tidyverse")

install.packages("stringr")
library("stringr")


# Make a table containing ONLY the samples that show detection of Ceratocystis fagacearum (samples 56, 77, 79, 80, 81, 82, 83, 84) and then order tables to show data in increasing order.

cFagacearumTbl <- notaxfill[,c(56, 77, 79:84, 94)]


# Make a table that is just sample 56 and order it in descending order

cFagaSamp56 <- notaxfill[, c(56,94)]
cFagaSamp56 <- arrange(cFagaSamp56, ... = -cFagaSamp56$OW.ITSF.2021.Plate1InsectRpd.B56, .by_group = FALSE)

cFagaSamp56Abun <- cFagaSamp56[str_detect(cFagaSamp56$taxonomy, "fagacearum"), ]

# ^ in this created dataframe, the row number corresponds to the abundance rank of C. fagacearum in that specific sample. Do this for the rest of the C. fagacearum-containing samples.



# Sample 77

cFagaSamp77 <- notaxfill[, c(77,94)]
cFagaSamp77 <- arrange(cFagaSamp77, ... = -cFagaSamp77$OW.ITSF.2021.Plate1InsectRpd.B77, .by_group = FALSE)

cFagaSamp77Abun <- cFagaSamp77[str_detect(cFagaSamp77$taxonomy, "fagacearum"), ]


# sample 79

cFagaSamp79 <- notaxfill[, c(79,94)]
cFagaSamp79 <- arrange(cFagaSamp79, ... = -cFagaSamp79$OW.ITSF.2021.Plate1InsectRpd.B79, .by_group = FALSE)

cFagaSamp79Abun <- cFagaSamp79[str_detect(cFagaSamp79$taxonomy, "fagacearum"), ]


# sample 80

cFagaSamp80 <- notaxfill[, c(80,94)]
cFagaSamp80 <- arrange(cFagaSamp80, ... = -cFagaSamp80$OW.ITSF.2021.Plate1InsectRpd.B80, .by_group = FALSE)

cFagaSamp80Abun <- cFagaSamp80[str_detect(cFagaSamp80$taxonomy, "fagacearum"), ]


## sample 81

cFagaSamp81 <- notaxfill[, c(81,94)]
cFagaSamp81 <- arrange(cFagaSamp81, ... = -cFagaSamp81$OW.ITSF.2021.Plate1InsectRpd.B81, .by_group = FALSE)

cFagaSamp81Abun <- cFagaSamp81[str_detect(cFagaSamp81$taxonomy, "fagacearum"), ]


# sample 82

cFagaSamp82 <- notaxfill[, c(82,94)]
cFagaSamp82 <- arrange(cFagaSamp82, ... = -cFagaSamp82$OW.ITSF.2021.Plate1InsectRpd.B82, .by_group = FALSE)

cFagaSamp82Abun <- cFagaSamp82[str_detect(cFagaSamp82$taxonomy, "fagacearum"), ]


# sample 83

cFagaSamp83 <- notaxfill[, c(83,94)]
cFagaSamp83 <- arrange(cFagaSamp83, ... = -cFagaSamp83$OW.ITSF.2021.Plate1InsectRpd.B83, .by_group = FALSE)

cFagaSamp83Abun <- cFagaSamp83[str_detect(cFagaSamp83$taxonomy, "fagacearum"), ]


# sample 84

cFagaSamp84 <- notaxfill[, c(84,94)]
cFagaSamp84 <- arrange(cFagaSamp84, ... = -cFagaSamp84$OW.ITSF.2021.Plate1InsectRpd.B84, .by_group = FALSE)

cFagaSamp84Abun <- cFagaSamp84[str_detect(cFagaSamp84$taxonomy, "fagacearum"), ]


# Make an excel file that has worksheets for each C. fagacearum-containing sample

install.packages("openxlsx")
library("openxlsx")

work_bookcFaga <- createWorkbook()
addWorksheet(work_bookcFaga, "sample56")
addWorksheet(work_bookcFaga, "sample77")
addWorksheet(work_bookcFaga, "sample79")
addWorksheet(work_bookcFaga, "sample80")
addWorksheet(work_bookcFaga, "sample81")
addWorksheet(work_bookcFaga, "sample82")
addWorksheet(work_bookcFaga, "sample83")
addWorksheet(work_bookcFaga, "sample84")

writeData(work_bookcFaga, "sample56", cFagaSamp56)
writeData(work_bookcFaga, "sample77", cFagaSamp77)
writeData(work_bookcFaga, "sample79", cFagaSamp79)
writeData(work_bookcFaga, "sample80", cFagaSamp80)
writeData(work_bookcFaga, "sample81", cFagaSamp81)
writeData(work_bookcFaga, "sample82", cFagaSamp82)
writeData(work_bookcFaga, "sample83", cFagaSamp83)
writeData(work_bookcFaga, "sample84", cFagaSamp84)

workbookPath <- paste("/isilon/cfia-ottawa-fallowfield/users/girouxeml/gladishd/pirl_working/oakwilt_redo/otu_tables")
dir.create(workbookPath)

saveWorkbook(work_bookcFaga, file = paste(workbookPath, "ordered_cFagacearum_containing_samples.xlsx", sep = "/"), overwrite = TRUE)
```
